File: skills/architecture/scripts/check-locality.mjs
import fs from 'fs';
import path from 'path';
import { extractImports } from './utils/extractor.mjs';
import { findCycles } from './utils/graph.mjs';

function walkDir(dir, exclude) {
    let files = [];
    try {
        const list = fs.readdirSync(dir);
        for (const file of list) {
            if (exclude.some(ex => file.includes(ex))) continue;
            const fullPath = path.join(dir, file);
            let stat;
            try {
                stat = fs.statSync(fullPath);
            } catch (e) {
                // Skip files/dirs we can't read (permissions, symlinks, etc.)
                continue;
            }
            if (stat.isDirectory()) {
                files = files.concat(walkDir(fullPath, exclude));
            } else if (fullPath.endsWith('.ts') || fullPath.endsWith('.tsx') || fullPath.endsWith('.js')) {
                files.push(fullPath);
            }
        }
    } catch (e) {
        // Skip directories we can't read
    }
    return files;
}

export function runLocalityCheck(targetDir, exclude = [
    'node_modules', '.test.', '.spec.',
    '.git', '.svn', '.hg',
    '.pytest_cache', '.tox', '__pycache__',
    '.venv', 'venv', '.env',
    'dist', 'build', 'coverage', '.coverage',
    '.next', '.nuxt', '.cache', '.parcel',
    'node_modules', '.npm', '.yarn',
    'target', '.gradle', '.m2',
    '.pytest', '.mypy_cache', '.ruff_cache',
    '.vscode', '.idea', '.DS_Store'
]) {
    const files = walkDir(targetDir, exclude);
    const graph = {};

    for (const file of files) {
        const content = fs.readFileSync(file, 'utf8');
        const imports = extractImports(content);
        graph[file] = [];
        
        for (const imp of imports) {
            // Only care about relative imports for locality
            if (imp.startsWith('.')) {
                // Crude resolution: just append .ts (assumes fixtures logic for now)
                // A robust script would need proper node module resolution here
                const resolved = path.resolve(path.dirname(file), imp);
                if (fs.existsSync(`${resolved  }.ts`)) {
                     graph[file].push(`${resolved  }.ts`);
                } else if (fs.existsSync(`${resolved  }.tsx`)) {
                     graph[file].push(`${resolved  }.tsx`);
                } else if (fs.existsSync(`${resolved  }/index.ts`)) {
                     graph[file].push(`${resolved  }/index.ts`);
                }
            }
        }
    }

    const cycles = findCycles(graph);
    
    // Calculate Fan-out
    const fanOut = Object.entries(graph)
        .map(([file, deps]) => ({ file, count: deps.length }))
        .sort((a, b) => b.count - a.count);

    return { cycles, fanOut };
}

// CLI entry point
if (process.argv[1] && (process.argv[1] === new URL(import.meta.url).pathname || process.argv[1].endsWith('check-locality.mjs'))) {
    const dir = process.argv[2] || 'src';
    console.log(`Checking locality in ${dir}...`);
    try {
      const { cycles, fanOut } = runLocalityCheck(path.resolve(process.cwd(), dir));
      console.log('\n--- Circular Dependencies ---');
      cycles.forEach((cycle, i) => {
          console.log(`\nCycle ${i + 1}:`);
          cycle.forEach(c => console.log(`  - ${path.relative(process.cwd(), c)}`));
      });
      if (cycles.length === 0) console.log('None found.');

      console.log('\n--- Top 5 Fan-out (Highest Imports) ---');
      fanOut.slice(0, 5).forEach(f => {
          console.log(`  - ${path.relative(process.cwd(), f.file)} (${f.count} imports)`);
      });
    } catch (e) {
      if(e.code === 'ENOENT') {
         console.error(`Directory not found: ${dir}`);
      } else {
         console.error(e);
      }
    }
}

File: skills/architecting/scripts/check_locality.py
import os
import sys
from typing import List, Optional, Tuple, Dict, Any
from utils.extractor import extract_imports, detect_lang
from utils.graph import find_cycles
from utils.walk import walk_dir, DEFAULT_EXCLUDE


def import_candidates(from_file: str, imp: str, lang: str) -> List[str]:
    from_dir = os.path.dirname(from_file)

    if lang == "py":
        # Leading dots are package levels: one dot = current package (same dir),
        # each extra dot = one parent up.
        dots = 0
        while dots < len(imp) and imp[dots] == ".":
            dots += 1

        if dots == 0:
            # Absolute import or not relative in the expected way
            # But locality check only cares about relative imports (starting with '.')
            return []

        rest = imp[dots:].split(".")
        rest = [r for r in rest if r]

        base = from_dir
        for _ in range(1, dots):
            base = os.path.dirname(base)

        resolved = os.path.join(base, *rest)
        # ponytail: bare "from . import x" (rest=[]) only resolves via __init__.py;
        # PEP 420 namespace packages without __init__.py won't match. Fix if that
        # pattern shows up: capture the imported names and try them as submodules.
        return [f"{resolved}.py", os.path.join(resolved, "__init__.py")]

    # js / go
    resolved = os.path.abspath(os.path.join(from_dir, imp))
    candidates = []

    basename = os.path.basename(imp)
    if "." in basename:
        # Import already carries an extension
        candidates.append(resolved)
        # TS allows a '.js' specifier to resolve to the '.ts' source.
        no_ext, _ = os.path.splitext(resolved)
        candidates.extend([f"{no_ext}.ts", f"{no_ext}.tsx"])

    candidates.extend(
        [
            f"{resolved}.ts",
            f"{resolved}.tsx",
            f"{resolved}.js",
            f"{resolved}.jsx",
            f"{resolved}.mjs",
            f"{resolved}.cjs",
            f"{resolved}.go",
            os.path.join(resolved, "index.ts"),
            os.path.join(resolved, "index.tsx"),
            os.path.join(resolved, "index.js"),
            os.path.join(resolved, "index.jsx"),
            os.path.join(resolved, "index.mjs"),
        ]
    )
    return candidates


def run_locality_check(
    target_dir: str, exclude: Optional[List[str]] = None
) -> Tuple[List[List[str]], List[Dict[str, Any]]]:
    if exclude is None:
        exclude = DEFAULT_EXCLUDE

    files = walk_dir(target_dir, exclude)
    graph = {}

    for file_path in files:
        try:
            with open(file_path, "r", encoding="utf-8") as f:
                content = f.read()
        except (UnicodeDecodeError, PermissionError):
            continue

        lang = detect_lang(file_path)
        imports = extract_imports(content, lang)
        graph[file_path] = []

        for imp in imports:
            # Only care about relative imports for locality
            if not imp.startswith("."):
                continue

            for candidate in import_candidates(file_path, imp, lang):
                if os.path.exists(candidate):
                    graph[file_path].append(candidate)
                    break

    cycles = find_cycles(graph)

    # Calculate Fan-out
    fan_out = []
    for file, deps in graph.items():
        fan_out.append({"file": file, "count": len(deps)})

    fan_out.sort(key=lambda x: x["count"], reverse=True)

    return cycles, fan_out


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(
        description="Check locality and circular dependencies."
    )
    parser.add_argument("dir", nargs="?", default="src", help="Directory to analyze")

    args = parser.parse_args()

    print(f"Checking locality in {args.dir}...")
    abs_dir = os.path.abspath(args.dir)

    try:
        cycles, fan_out = run_locality_check(abs_dir)

        print("\n--- Circular Dependencies ---")
        if not cycles:
            print("None found.")
        else:
            for i, cycle in enumerate(cycles):
                print(f"\nCycle {i + 1}:")
                for node in cycle:
                    print(f"  - {os.path.relpath(node, os.getcwd())}")

        print("\n--- Top 5 Fan-out (Highest Imports) ---")
        for item in fan_out[:5]:
            print(
                f"  - {os.path.relpath(item['file'], os.getcwd())} ({item['count']} imports)"
            )

    except FileNotFoundError:
        print(f"Directory not found: {args.dir}", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"An error occurred: {e}", file=sys.stderr)
        sys.exit(1)
